pkg/apiserver/server.go

```diff
@@ -70,20 +70,25 @@ func newEngine() *gin.Engine {
 		v.RegisterValidation("RequiredFieldEqual", apivalidator.RequiredFieldEqualValid, true)
 	}
 
-	moveToUIRoot := func(c *gin.Context) {
-		c.Redirect(http.StatusMovedPermanently, "/dashboard")
-	}
-
-	r.GET("/", moveToUIRoot)
 	ui := uiserver.AssetsFS()
 	if ui != nil {
-		newDashboardRouter(r, ui)
+		r.GET("/", func(c *gin.Context) {
+			c.FileFromFS("/", ui)
+		})
+		r.GET("/:foo/*bar", func(c *gin.Context) {
+			c.FileFromFS("/", ui)
+		})
```

Comment:
what does `/:foo/*bar` means? I do not found and reference in gin's doc.

Reply:
> I do not found and reference in gin's doc.

It still lacks docs.

An example, `/experiments` will match `/:foo` and `experiments/xxx` will match `/:foo/*bar`.

Because gin does not permit to define root route wildcard (`/*any`) and before https://github.com/gin-gonic/gin/pull/2663 (This include in gin `1.7`) it's also not permitted to define mixed param and non-param paths (Like `/:foo ` and `/api`).

So with the latest gin, I can use `/:foo/*bar` to simulate all of the frontend routes. Without the changes above, we can only define the **exact** routes like below:

- `/experiments`
- `/experiments/:uid`
- `/schedules`
- and so on

This would be very redundant.

In the past, we have been limited by the [httprouter](https://github.com/julienschmidt/httprouter), you can found the history reason in https://github.com/julienschmidt/httprouter/pull/329, This pr still not merged. Roughly speaking, httprouter forgoes fuzzy matching for the sake of speed.

Reply:
Oh, I get that, the `:foo` and `*bar` are only placeholders, not meaningful parameters.
This route will show `index.html` on all the paths (meaningful in frontend), like `/expriments/<uid>`, then let javascript parse/resolve the path for showing certain pages.

Could you append some comments on the `/:foo/*bar`, the name is ambiguous.

Reply:
Got it. I will add some comments later.